SQRLDotNetClientUI/ViewModels/IdentitySettingsViewModel.cs

```diff
@@ -0,0 +1,121 @@
+        public string ProgressText { get => _progressText; set => this.RaiseAndSetIfChanged(ref _progressText, value); }
+
+        public IdentitySettingsViewModel() { }
+
+        public IdentitySettingsViewModel(SQRL sqrlInstance, SQRLIdentity identity)
+        {
+            this.Title = "SQRL Client - Identity Settings";
+            this.SqrlInstance = sqrlInstance;
+            this.Identity = identity;
+            this.IdentityCopy = identity.Clone();
+
+            if (identity != null) this.Title += " (" + identity.IdentityName + ")";
+        }
+
+        public void Close()
+        {
+            ((MainWindowViewModel)AvaloniaLocator.Current.GetService<MainWindow>().DataContext).Content =
+                ((MainWindowViewModel)AvaloniaLocator.Current.GetService<MainWindow>().DataContext).MainMenu;
+        }
+
+        public async void Save()
+        {
+            CanSave = false;
+
+            if (!HasChanges())
+            {
+                Close();
+                CanSave = true;
+                return;
+            }
+
+            string password = "test12345678";
+
+            this.ProgressMax = (double)Identity.Block1.IterationCount;
+
+            var progress = new Progress<KeyValuePair<int, string>>(progress =>
+            {
+                this.ProgressPercentage = (double)progress.Key;
+                this.ProgressText = progress.Value + progress.Key;
+            });
+
+            (bool ok, byte[] imk, byte[] ilk) = await SqrlInstance.DecryptBlock1(Identity, password, progress);
+
+            if (!ok)
+            {
+                var messageBoxStandardWindow = MessageBox.Avalonia.MessageBoxManager.GetMessageBoxStandardWindow(
+                    $"Error", $"The identity could not be decrypted using the given password! Please try again!", 
+                    MessageBox.Avalonia.Enums.ButtonEnum.Ok, 
+                    MessageBox.Avalonia.Enums.Icon.Error);
+
+                await messageBoxStandardWindow.ShowDialog(AvaloniaLocator.Current.GetService<MainWindow>());
+
+                CanSave = true;
+                return;
+            }
+
+            this.ProgressMax = (double)IdentityCopy.Block1.IterationCount;
+
+            SQRLIdentity id = await SqrlInstance.GenerateIdentityBlock1(
+                imk, ilk, password, IdentityCopy, progress, IdentityCopy.Block1.PwdVerifySeconds);
+
+            // Swap out the old type 1 block with the updated one
+            // TODO: We should probably make sure that this is an atomic operation
+            Identity.Blocks.Remove(Identity.Block1);
+            Identity.Blocks.Insert(0, id.Block1);
+
+            CanSave = true;
+            Close();
+        }
+
+        /// <summary>
+        /// Returns <c>true</c> if any of the identity settings were changed by the 
+        /// user are and those changes have not been applied yet, or <c>false</c> otherwise.
+        /// </summary>
+        public bool HasChanges()
```

Comment:
Note sure if you saw this or not, thoughts? @alexhauser

Reply:
Your neither wrong nor crazy, @josegomez! 😁

I needed to check for `HintLength`, `PwdTimeoutMins` and `PwdVerifySeconds` separately, because they're not part of the option flags, but all the others could be checked by just comparing the `FlagsValue`. Thanks for the hint, I will post an update!

Reply:
👍